controllers/core_collector_controller.go

```diff
@@ -70,14 +76,62 @@ func (r *CoreCollectorReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	})
 	r.Client = mgr.GetClient()
 	r.scheme = mgr.GetScheme()
+
+	periodicRecChan := make(chan event.GenericEvent, 5)
+
+	go func() {
+		ticker := time.NewTicker(1 * time.Minute)
+		defer ticker.Stop()
+		for range ticker.C {
+			mockObj := &opnimonitoringv1beta1.CollectorConfig{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      otel.MetricsCrdName,
+					Namespace: "fake",
+				},
+			}
+			periodicRecChan <- event.GenericEvent{
+				Object: mockObj,
+			}
+		}
+	}()
+
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&corev1beta1.Collector{}).
 		Watches(&source.Kind{Type: &opnimonitoringv1beta1.CollectorConfig{}}, requestMapper).
 		Watches(&source.Kind{Type: &opniloggingv1beta1.CollectorConfig{}}, requestMapper).
-		Watches(&source.Kind{Type: &promoperatorv1.ServiceMonitor{}}, requestMapper).
-		Watches(&source.Kind{Type: &promoperatorv1.PodMonitor{}}, requestMapper).
-		Watches(&source.Kind{Type: &corev1.Pod{}}, requestMapper).
-		Watches(&source.Kind{Type: &corev1.Service{}}, requestMapper).
+		// for metrics, the we want to watch changes to the spec of objects that drive discovery
+		Watches(
+			&source.Kind{Type: &promoperatorv1.ServiceMonitor{}},
+			requestMapper,
+			builder.WithPredicates(
+				predicate.Or(predicate.GenerationChangedPredicate{}),
+			),
+		).
+		Watches(
+			&source.Kind{Type: &promoperatorv1.PodMonitor{}},
+			requestMapper,
+			builder.WithPredicates(
+				predicate.Or(predicate.GenerationChangedPredicate{}),
+			),
+		).
+		Watches(
+			&source.Kind{Type: &corev1.Service{}},
+			requestMapper,
+			builder.WithPredicates(
+				predicate.Or(
+					predicate.GenerationChangedPredicate{},
+				),
+			),
+		).
+		// We need to watch almost all pod events here, but instead of queuing reconciles
+		// for each event, we "hack" a batch mechanism for potential pod events
+		// using a periodic reconciliation channel.
+		Watches(
+			&source.Channel{
+				Source: periodicRecChan,
+			},
+			requestMapper,
+		).
```

Comment:
As discussed on slack a better mechanism is to use the `RequeueAfter` field of `reconcile.Result` to force this to requeue after 60 seconds.

A couple of alternative approaches might also be:
- Use event filtering to filter out some of the pod update events
- Use a ratelimiter that's more aggressive that the 100, 10qps default token bucket.

These can be added to the controller with the `WithEventFilter` and `WithOptions` functions.

Reply:
Seems `WithEventFilter` is equivalent to (globally) setting predicates on watched resources.

I tried using the following predicates on pod events, but for whatever reason it wasn't catching updates to pods that were managed by other controllers

```
		Watches(
			&source.Kind{Type: &corev1.Pod{}},
			requestMapper,
			builder.WithPredicates(
				predicate.Or(
					predicate.GenerationChangedPredicate{},
					predicate.ResourceVersionChangedPredicate{},
					predicate.NewPredicateFuncs(
						func(obj client.Object) bool {
							return true
						},
					),
				),
			),
		).		
```

My fallback is to re-queue every minute when metrics is enabled, but if we can find a set of predicates that work on pods we can probably bump up that re-queue timer to 15minutes, as a safeguard to catch updates